botocore/endpoint.py

```diff
@@ -216,9 +216,34 @@ def _do_get_response(self, request, operation_model):
         parser = self._response_parser_factory.create_parser(protocol)
         parsed_response = parser.parse(
             response_dict, operation_model.output_shape)
+        # Do a second parsing pass to pick up on any modeled error fields
+        # NOTE: Ideally, we would push this down into the parser classes but
+        # they currently have no reference to the operation or service model
+        # The parsers should probably take the operation model instead of
+        # output shape but we can't change that now
+        if http_response.status_code >= 300:
+            self._add_modeled_error_fields(
+                response_dict, parsed_response,
+                operation_model, parser,
+            )
         history_recorder.record('PARSED_RESPONSE', parsed_response)
         return (http_response, parsed_response), None
 
+    def _add_modeled_error_fields(
+            self, response_dict, parsed_response,
+            operation_model, parser,
+    ):
+        error_code = parsed_response.get("Error", {}).get("Code")
+        if error_code is None:
+            return
+        service_model = operation_model.service_model
+        error_shape = service_model.shape_for_error_code(error_code)
+        if error_shape is None:
+            return
+        modeled_parse = parser.parse(response_dict, error_shape)
+        # TODO: avoid naming conflicts with ResponseMetadata and Error
```

Comment:
Did we at least add a test here to detect when the collisions occur?

Reply:
I've updated with some tests to validate this. It's unfortunately a little trickier than it seems at first glance to add this. Because `Error` and `ResponseMetadata` are injected at the parser level we need to do the renaming there, ideally doing the rename when we're parsing the members for a structure (in the `_handle_structure` methods of the parsers). The unfortunate thing is we don't have a great way to know when we're parsing a top-level structure vs a nested structure which we need for `ResponseMetadata` renames. We can work around this for `Error` as that only applies to structures with the trait `exception`. For `ResponseMetadata` we might have to just audit everywhere we might inject that key and ensure we aren't clobbering the field if it doesn't already exist.